prow/plugins/mergecommitblocker/mergecommitblocker.go

```diff
@@ -0,0 +1,113 @@
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/labels"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	// PluginName defines this plugin's registered name.
+	PluginName = "mergecommitblocker"
+)
+
+// init registers out plugin as a pull request handler
+func init() {
+	plugins.RegisterPullRequestHandler(PluginName, handlePullRequest, helpProvider)
+}
+
+// helpProvider provides information on the plugin
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	// Only the Description field is specified because this plugin is not triggered with commands and is not configurable.
+	return &pluginhelp.PluginHelp{
+		Description: fmt.Sprintf(`The merge commit blocker plugin adds the do-not-merge/contains-merge-commit label to pull requests that contain merge commits`),
+	}, nil
+}
+
+type githubClient interface {
+	AddLabel(org, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+}
+
+func handlePullRequest(pc plugins.Agent, pre github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionOpened &&
+		pre.Action != github.PullRequestActionReopened &&
+		pre.Action != github.PullRequestActionSynchronize {
+		return nil
+	}
+	return handle(pc.GitHubClient, pc.GitClient, pc.Logger, &pre)
+}
+
+func handle(ghc githubClient, gc *git.Client, log *logrus.Entry, pre *github.PullRequestEvent) error {
+	var (
+		owner = pre.PullRequest.Base.Repo.Owner.Login
+		repo  = pre.PullRequest.Base.Repo.Name
+		num   = pre.PullRequest.Number
+	)
+
+	// Clone the repo, checkout the PR.
+	r, err := gc.Clone(fmt.Sprintf("%s/%s", owner, repo))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := r.Clean(); err != nil {
+			log.WithError(err).Error("Error cleaning up repo.")
+		}
+	}()
+	if err := r.CheckoutPullRequest(num); err != nil {
+		return err
+	}
+	target, head := "origin/master", "HEAD"
+	// If we have specific SHAs, use them.
+	if pre.PullRequest.Base.SHA != "" {
+		target = pre.PullRequest.Base.SHA
+	}
+	if pre.PullRequest.Head.SHA != "" {
+		head = pre.PullRequest.Head.SHA
+		if err := r.Checkout(pre.PullRequest.Head.SHA); err != nil {
+			return err
+		}
+	}
+
+	existMergeCommits, err := r.MergeCommitsExistBetween(target, head)
+	if err != nil {
+		return err
+	}
+	issueLabels, err := ghc.GetIssueLabels(owner, repo, num)
+	if err != nil {
+		return err
+	}
+	hasLabel := github.HasLabel(labels.MergeCommits, issueLabels)
+	if hasLabel && !existMergeCommits {
+		log.Infof("Removing %q Label for %s/%s#%d", labels.MergeCommits, owner, repo, num)
+		return ghc.RemoveLabel(owner, repo, num, labels.MergeCommits)
+	} else if !hasLabel && existMergeCommits {
+		log.Infof("Adding %q Label for %s/%s#%d", labels.MergeCommits, owner, repo, num)
+		return ghc.AddLabel(owner, repo, num, labels.MergeCommits)
```

Comment:
Do we want to add a comment here to explain why the label is being added?

/cc @kubernetes/sig-contributor-experience-pr-reviews

Reply:
Added a comment.

Reply:
I think what @stevekuznetsov was asking was should we add a comment to the issue itself, rather than a code comment.

An example would be the `blockade` plugin, that sets a comment on the issue when it adds the label, and prunes the comments when they no longer apply (https://github.com/kubernetes/test-infra/blob/ea3dad5c1be6a4b7e4cff6283519286e6adeaabb/prow/plugins/blockade/blockade.go#L162-L178)

Reply:
Oops, hard to follow the "slang" here...

Updated. PTAL. Feel free to propose changes on "fixme" message.